test/test_cnf.py

```diff
@@ -59,3 +64,32 @@ def test_normalize_einsum(equation, plates, backend, einsum_impl):
 
     actual = eval(quote(expected))  # requires torch, bint
     assert_close(actual, expected)
+
+
+@pytest.mark.parametrize("x_shape", [(), (1,), (3,), (1, 1), (1, 3), (2, 1), (2, 3)], ids=str)
+@pytest.mark.parametrize("y_shape", [(), (1,), (3,), (1, 1), (1, 3), (2, 1), (2, 3)], ids=str)
+@pytest.mark.parametrize("x_inputs,y_inputs", [
+    ("", ""),
+    ("i", ""),
+    ("", "i"),
+    ("j", "i"),
+    ("ij", "i"),
+    ("i", "ik"),
+    ("ij", "ik"),
+])
+@pytest.mark.parametrize("red_op,bin_op", [(ops.add, ops.mul), (ops.logaddexp, ops.add)], ids=str)
```

Comment:
Add `(ops.max, ops.add)` for `torch_map` here?

Reply:
Pyro doesn't actually have a (max,add) ring, rather it has an (argmax,add) ring that is tightly coupled to our pyro adjoint algorithms including requirements for `._pyro_dims` attribute. While we could work around this with some effort (effort I could spend elsewhere), the (argmax,add) ring simply adds-then-reduces under the hood, so there is no computational advantage.

Reply:
> Pyro doesn't actually have a (max,add) ring

The forward pass looks to me like [it computes `max`](https://github.com/pyro-ppl/pyro/blob/dev/pyro/ops/einsum/torch_map.py#L37), but as you say there's no computational advantage and the adjoint behavior is different.